### DISCUSSION TITLE
Add an async version of `.send()`

### DISCUSSION THREAD

**Discussion by hw2zheng** [statelyai/xstate#4120] [General]
When sending an event to an xstate machine, the `.send()` function returns when the machine has reached the end state for that event. This works pretty well when the state machine doesn't rely on any async functions, but it returns prematurely if a state invokes an async function. 

In the example below, the `test` event causes the machine to transition from the `start` state to the `invokeFn` state. The `testService.send` function returns the state at that point. However, when the async function resolves the machine will transition to the `done` state with no new user input, so it doesn't feel like `invokeFn` is the "end state" for that event.

Is it possible to add something like `machine.sendAsync()`, which does the same thing as `.send()` but lets any invocations finish before resolving? Example and workaround below: 

```ts
const machine = createMachine({
  predictableActionArguments: true,
  initial: "start",
  states: {
    start: {
      on: {
        test: {
          target: "invokeFn"
        }
      }
    },
    invokeFn: {
      invoke: {
        src: async () => {
          return "test data";
        },
        onDone: {
          target: "done"
        }
      }
    },
    done: {
      type: "final"
    }
  }
});

const testService = interpret(machine);
testService.start();
const result = testService.send("test");
console.log(result.value);
```
[sandbox](https://codesandbox.io/s/zealous-water-zy8ntn?file=/src/index.js:262-800)

For the moment I'm using `onTransition` and `nextEvents` to explicitly wait for all invocations to finish before returning. It works for my use case, but it feels kinda hacky
```
testService.onTransition((state) => {
  if (!state.nextEvents.find((x) => x.includes(":invocation"))) {
    console.log(state.value);
  }
});
```

**Comment by ChrisShank** [comment]
You are probably interesting in the `waitFor` helper https://xstate.js.org/docs/guides/interpretation.html#waitfor
